### DISCUSSION TITLE
How to handle multiple errors in (react query) using (hono rpc)?

### DISCUSSION THREAD

**Discussion by igmtink** [honojs/hono#2753] [Q&A] [answered]
How can I handle multiple errors like in my case?

![image-1](https://github.com/honojs/hono/assets/117252369/b8f4a532-7518-4a9b-93e8-6c49c551cae9)
![image-2](https://github.com/honojs/hono/assets/117252369/982839c0-9317-403f-a2df-377453bfc920)

What I want is to display the error coming from the backend honojs, to the client side for the best ux experience, displayig the `Username is already exist` or `Email is already exist` when the user is registering.

**Comment by yusukebe** [comment]
Hi @igmtink 

Try to write `if (!response.ok)`  first.

```ts
if (!response.ok) {
  const data = await response.json()
}

// ...
```

**Reply by igmtdev** [reply]
thank you sir, that's a good idea, returning `await response.json()` so it will return the `error` inside of `!response.ok`

**Reply by yusukebe** [reply]
@igmtdev 

If you are okay, please mark it as the answer!

![CleanShot 2024-05-22 at 17 32 35@2x](https://github.com/honojs/hono/assets/10682/8baa473c-1415-4b8a-9a34-036553354dc5)

**Reply by igmtdev** [reply]
Good day sir @yusukebe 

Still have a problem sir, the type of data is `{ error: string } | { data: string }` but I can't get the message from the `data.error`

I want to `throw new Error(data.error)` or `throw new Error(data)` so I can get the `error message` using the `onError` of `useMutation`

![image](https://github.com/honojs/hono/assets/161505752/dc25699e-7555-4d0d-8213-4ffa08cf2fe5)
![image](https://github.com/honojs/hono/assets/161505752/fbf1df9f-0763-4bbb-8970-ebdd3285c595)

Here's the setup of my sign up route, also I have 3 method here to handle the error because I don't know what is the correct to handle the error to pass to client side.
![image](https://github.com/honojs/hono/assets/161505752/f1b0ca40-095d-4ec6-9b82-8732350ae819)

And for the index of my hono
![image](https://github.com/honojs/hono/assets/161505752/5ba48dcf-5ce0-4ef0-af18-4943b2d0cad2)

**Reply by yusukebe** [reply]
@igmtdev 

Have you ever tried to use the status code?

```ts
if (response.status === 409) {
  const data = await response.json()
}
```

**Reply by igmtdev** [reply]
I fix it already sir, so the solution here is to move inside of if statement, I can't understand what's happening, I can get the `data.error`, but in outside I can't get the `data.error` or the `data.data`?

```typescript
if (!response.ok) {
  const data = await response.json()
  throw new Error(data.error)
}
```

![image](https://github.com/honojs/hono/assets/161505752/e44bf16f-1f09-47e1-acbc-a97e204d3a3e)

**Reply by yusukebe** [reply]
Ah. With the current implementation and spec, you should write `if(!response.ok) {}` first. Because the response type is like the following:

![CleanShot 2024-05-22 at 22 07 23@2x](https://github.com/honojs/hono/assets/10682/9e0fa9f1-73e6-4e28-97f6-674f4f31fb76)

Hey @kosei28. Do you have any thoughts about this matter?

**Reply by igmtdev** [reply]
Yes sir I understand now that I need to write the `if (!response.ok)` because the response type is it's either `{ error: string }` or `{ data: string }`, but why I can't destructure the `{ data }` when I'm done to write the `if (!response.ok)`

I just want to destructure both `data` and the `error` so it would better to read the code like this:

![image](https://github.com/honojs/hono/assets/161505752/4aba550d-7da7-4bd9-a5dc-7eb2788a05c3)

but it's seems there's a type error:

![image](https://github.com/honojs/hono/assets/161505752/975cb4a5-4082-425f-b0f9-ad077991b73b)

**Reply by igmtdev** [reply]
It's seems I can't destructure the `const { data } = await response.json()` to return only the `data` if the `response` is `ok`
But the I can destructure the `const { error } = await response.json()` inside of `if (!response.ok)` to return only the error

But if I do this, there's no type error, by returning `await response.json()`

![image](https://github.com/honojs/hono/assets/161505752/7067b501-8553-46d6-a82e-896a1aecd094)

**Reply by kosei28** [reply]
You seem to be confusing `const data = await response.json()` with `const { data } = await response.json()`. Note that they are different codes where the object and its properties are assigned.

**Reply by yusukebe** [reply]
Hmm. Now, it's hard to debug it well with just this information. If you share the minimal project to reproduce it, I or anyway can investigate.

Anyway, TypeScript is fun but difficult.

**Reply by yusukebe** [reply]
@kosei28 I see. Thanks!

**Reply by igmtdev** [reply]
sir @kosei28 why I can't use this `const { data } = await response.json()` to return the `data` even I handle the `if (!response.ok)`, inside of `if (!response.ok)` I can destructure the `error` from `const { error } = await response.json()`?

**Reply by igmtdev** [reply]
[Here's my project, it's a minimal project didn't start yet](https://github.com/igmtdev/rpc-type-test/blob/main/hooks/features/api/users/use-sign-up.ts)

**Reply by igmtdev** [reply]
I just want my code to be like this for better to read, but I can't destructure the `{ data }`

![image](https://github.com/honojs/hono/assets/161505752/2d20a39e-dfd1-4972-98d1-1e45b157d34f)

**Reply by kosei28** [reply]
It seems that Hono's ResponseType is used for useMutation, but in this case it becomes `{ data: string } | { error: string }`. `const { data } = await response.json()`'s `data` is of string type, so a type error will occur.

**Reply by igmtdev** [reply]
so sir @kosei28 there's nothing I can do here? I just will `return await response.json()`?

**Reply by kosei28** [reply, marked as answer]
I think the type error will be resolved by changing ResponseType as follows.
```ts
type ResponseType = InferResponseType<typeof client.api.signup.$post, 200>['data']
```

**Reply by igmtdev** [reply]
thank you sir @kosei28 it works :)
